src/libs/kata-types/src/dmverity.rs

```diff
@@ -0,0 +1,509 @@
+                if self.hash.len() != 40 || hex::decode(&self.hash).is_err() {
+                    bail!(
+                        "Invalid hash value sha1:{} for DmVerity device with sha1",
+                        self.hash,
+                    );
+                }
+            }
+            // TODO: support ["sha224", "sha384", "sha512", "ripemd160"];
+            _ => {
+                bail!(
+                    "Unsupported hash algorithm {} for DmVerity device {}",
+                    self.hashtype,
+                    self.hash,
+                );
+            }
+        }
+
+        if self.blocknum == 0 || self.blocknum > u32::MAX as u64 {
+            bail!("Zero block count for DmVerity device {}", self.hash);
+        }
+        if !Self::is_valid_block_size(self.blocksize) || !Self::is_valid_block_size(self.hashsize) {
+            bail!(
+                "Unsupported verity block size: data_block_size = {},hash_block_size = {}",
+                self.blocksize,
+                self.hashsize
+            );
+        }
+        if self.offset % self.hashsize != 0 || self.offset < self.blocksize * self.blocknum {
+            bail!(
+                "Invalid hashvalue offset {} for DmVerity device {}",
+                self.offset,
+                self.hash
+            );
+        }
+
+        Ok(())
+    }
+
+    fn is_valid_block_size(block_size: u64) -> bool {
+        for order in 9..20 {
+            if block_size == 1 << order {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+// Parse `DmVerityInfo` object from plaintext or base64 encoded json string.
+impl TryFrom<&str> for DmVerityInfo {
+    type Error = Error;
+
+    fn try_from(value: &str) -> std::result::Result<Self, Self::Error> {
+        let option = if let Ok(v) = serde_json::from_str::<DmVerityInfo>(value) {
+            v
+        } else {
+            let decoded = base64::decode(value)?;
+            serde_json::from_slice::<DmVerityInfo>(&decoded)?
+        };
+
+        option.validate()?;
+        Ok(option)
+    }
+}
+
+impl TryFrom<&String> for DmVerityInfo {
+    type Error = Error;
+
+    fn try_from(value: &String) -> std::result::Result<Self, Self::Error> {
+        Self::try_from(value.as_str())
+    }
+}
+
+/// Creates a mapping with <name> backed by data_device <source_device_path>
+/// and using hash_device for in-kernel verification.
+/// It will return the verity block device Path "/dev/mapper/<name>"
+/// Notes: the data device and the hash device are the same one.
+pub fn create_verity_device(
+    verity_option: &DmVerityInfo,
+    source_device_path: &Path,
+) -> Result<String> {
+    let dm = DM::new()?;
+    let verity_name = DmName::new(&verity_option.hash)?;
+    let id = DevId::Name(verity_name);
+    let opts = DmOptions::default().set_flags(DmFlags::DM_READONLY);
+    let hash_start_block: u64 =
+        (verity_option.offset + verity_option.hashsize - 1) / verity_option.hashsize;
```

Comment:
What if `verity_option.hashsize == 0`?

Reply:
Fixed. `hashsize` would be validated in `validate()` before we use it. 
```
if !Self::is_valid_block_size(self.blocksize) || !Self::is_valid_block_size(self.hashsize) {
            bail!(
                "Unsupported verity block size: data_block_size = {},hash_block_size = {}",
                self.blocksize,
                self.hashsize
            );
        }
```